### DISCUSSION TITLE
Downloading Latest release via C#

### DISCUSSION THREAD

**Discussion by IMTSS** [community/community#58810] [Programming Help]
### Body

Hi everyone,

I Have a question. How do I download the latest release of a repository via C#. I have created a Token (Classic) for my private repository, I selected the full control option for repository. My code does find the url to download the file but HttpClient keeps on returning 404 even though I send it my token. Can you please help. Im new to github and would like to use it as an update repository.. Below is my code : 

public UpdateChecker(string currentVersion)
        {
            this.currentVersion = currentVersion;
        }

        public async Task CheckForUpdates(GitHubClient gitHubClient)
        {
            try
            {
                var latestRelease = await gitHubClient.Repository.Release.GetLatest("IMTSS", "Test");
                string latestVersion = latestRelease.TagName;

                if (IsUpdateAvailable(latestVersion))
                {
                    Console.WriteLine("An update is available! Version " + latestVersion);
                    await AskUserToUpdate(latestRelease.Assets[0].BrowserDownloadUrl);
                }
                else
                {
                    Console.WriteLine("You are already using the latest version.");
                }
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error retrieving latest version from GitHub: " + ex.Message);
            }
        }

        private bool IsUpdateAvailable(string latestVersion)
        {
            if (!string.IsNullOrEmpty(latestVersion))
            {
                Version current = new Version(currentVersion);
                Version latest = new Version(latestVersion);
                return latest > current;
            }
            return false;
        }

        private async Task AskUserToUpdate(string downloadUrl)
        {
            
            DialogResult response = new DialogResult();
            response = MessageBox.Show("Would you like to update? (y/n)", "Confirm Download", MessageBoxButtons.YesNo);
            if (response == DialogResult.Yes)
            {
                await DownloadAndUpdate(downloadUrl);
            }
        }

        private async Task DownloadAndUpdate(string downloadUrl)
        {
            try
            {
               
                string tempFilePath = Path.GetTempFileName();
                using (HttpClient client = new HttpClient())
                {
                    client.DefaultRequestHeaders.Add("Authorization", string.Concat("token " , AccessToken));
                    var response = client.GetAsync(downloadUrl).Result;
                    response.EnsureSuccessStatusCode();
                    using (var stream = await response.Content.ReadAsStreamAsync())
                    {
                        using (var fileStream = File.Create(tempFilePath))
                        {
                            stream.CopyTo(fileStream);
                        }
                    }
                }
                string updateDirectory = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "Update");
                Directory.CreateDirectory(updateDirectory);
                string updateFile = Path.Combine(updateDirectory, "update.zip");
                File.Move(tempFilePath, updateFile);
                ExtractUpdateFiles(updateFile);
                CleanUp(updateFile);
                Console.WriteLine("Update complete! Please restart the application.");
            }
            catch (Exception ex)
            {
                Console.WriteLine("Error downloading or applying update: " + ex.Message);
            }
        }

### Guidelines

- [X] I have read and understood this category's [guidelines](https://github.com/orgs/community/discussions/51384) before making this post.
EDIT: If I paste the "downloadUrl" in my browser that I'm logged in with it downloads the file.

**Comment by oliveirasmatheus** [comment]
Hi! It looks like you're on the right track with your code. The issue might be related to how the `HttpClient` is handling the `Authorization` header for private repositories. Here are a few suggestions to help you resolve the 404 issue:

1. **Ensure the Token Scope is Correct**:  
   Double-check that the personal access token (PAT) you generated includes the necessary repository access scopes. For private repositories, you need at least the `repo` scope for full control.

2. **Debug the Download URL**:  
   You mentioned that pasting the `downloadUrl` in your browser works, which suggests the URL is correct. Make sure the URL is accessible with the same credentials you’re using in your code.

3. **Fix the Authorization Header**:  
   Update your `Authorization` header format:
   ```csharp
   client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", AccessToken);
   ```
   Ensure `AccessToken` contains your PAT.

4. **Use `await` Consistently**:  
   Replace this line:  
   ```csharp
   var response = client.GetAsync(downloadUrl).Result;
   ```  
   With this:  
   ```csharp
   var response = await client.GetAsync(downloadUrl);
   ```  
   Blocking calls like `.Result` can lead to deadlocks or other unexpected issues in asynchronous methods.

5. **Check for Redirects**:  
   Sometimes, the `downloadUrl` might redirect. Ensure `HttpClient` follows redirects automatically:
   ```csharp
   client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Bearer", AccessToken);
   var response = await client.GetAsync(downloadUrl);
   response.EnsureSuccessStatusCode();
   ```

6. **Add Logging**:  
   Log the response status code and headers for better debugging:  
   ```csharp
   Console.WriteLine($"Response: {response.StatusCode}");
   Console.WriteLine($"Headers: {response.Headers}");
   ```

If you’ve tried the above steps and the issue persists, the problem might lie in how the `downloadUrl` is being generated or used in the `HttpClient` call. Let me know if you need further help!
